mm/ubsan/ubsan.c

```diff
@@ -158,9 +158,11 @@ static int64_t get_signed_val(FAR struct type_descriptor *type,
   if (is_inline_int(type))
     {
       unsigned extra_bits = sizeof(int64_t) * 8 - type_bit_width(type);
-      uintptr_t ulong_val = (uintptr_t)val;
+      uint64_t mask = (1llu << extra_bits) - 1;
+      uint64_t ret = (uint64_t)val & mask;
 
-      return ((int64_t)ulong_val) << extra_bits >> extra_bits;
+      return (int64_t)(((ret & (1llu << (extra_bits - 1))) != 0) ?
+             ret | ~mask : ret);
```

Comment:
I re-read my code again and it is a bit faulty. I mean that it will work for 32 bit values, but other variants seems to be wrong. It should be:
```
unsigned bits = type_bit_width(type);
uint64_t mask = (1llu << bits) - 1;
uint64_t ret = (uint64_t)val & mask;

return (int64_t)(((ret & (1llu << (bits - 1))) != 0) ?
             ret | ~mask : ret);
```

Reply:
I mean that:
1. 16 bits encoded value: `val` contain `0xFFFF` (`-1` in `int16_t`) and `type_bit_width(type)` is 16. Then `mask` will be `0xFFFF`, `ret` will be `0xFFFF`, `((ret & (1llu << (bits - 1))) != 0)` will be `0xFFFF & 0x8000 != 0` will be `true` and `0xFFFF | 0xFFFFFFFFFFFF0000` == `0xFFFFFFFFFFFFFFFF` (`-1` in `int64_t`) will be returned.
2. 32 bits encoded value: `val` contain `0xFFFFFFFF` (`-1` in `int32_t`) and `type_bit_width(type)` is 32. Then `mask` will be `0xFFFFFFFF`, `ret` will be `0xFFFFFFFF`, `((ret & (1llu << (bits - 1))) != 0)` will be `0xFFFFFFFF & 0x80000000 != 0` will be `true` and `0xFFFFFFFF | 0xFFFFFFFF00000000` == `0xFFFFFFFFFFFFFFFF` (`-1` in `int64_t`) will be returned.
3. 64 bits encoded value: `val` contain `0xFFFFFFFFFFFFFFFF` (`-1` in `int64_t`) and `type_bit_width(type)` is 64. Then `mask` will be `0xFFFFFFFFFFFFFFFF`, `ret` will be `0xFFFFFFFFFFFFFFFF`, `((ret & (1llu << (bits - 1))) != 0)` will be `0xFFFFFFFFFFFFFFFF & 0x8000000000000000 != 0` will be `true` and `0xFFFFFFFFFFFFFFFF | 0x0` == `0xFFFFFFFFFFFFFFFF` (`-1` in `int64_t`) will be returned.

Reply:
@pkarashchenko  I tried your new code, but in practice I don't see any difference. Both yield correct results.

As I see, no matter the original type of the value, it is always encoded as a 64-bit value.  
For example, consider the following:

```
int16_t a = -1;
int16_t b = a << 2;
```

The above will trigger UBSan, but the variable `val` in function `get_signed_val()` will contain the value `0xFFFFFFFFFFFFFFFF`.

*(As said, I can only test on simulator / 64-bit)*

I trust you on this, and I will create a new PR with the suggested fix, but in practice both work, at least for me.

Reply:
Maybe there is a way to test simulation in 32bit docker environment and see if that brings a difference

Reply:
@pkarashchenko Originally I was using the Lua tests suite. Some tests (I guess purposely) were triggering UB's on the VM itself.

Then, after your comments, I also added the following:

```c
uint32_t a = -1;
uint16_t b = -1;
uint8_t c = -1;

uint32_t k = a << 2;
uint16_t l = b << 2;
uint8_t m = c << 2;
```